src/libstd/sys/unix/process/process_unix.rs

```diff
@@ -35,17 +36,16 @@ impl Command {
         // accessing the `environ` pointer ourselves). Make sure no other thread
         // is accessing the environment when we do the fork itself.
         //
-        // Note that as soon as we're done with the fork there's no need to hold
-        // a lock any more because the parent won't do anything and the child is
-        // in its own process.
-        let result = unsafe {
-            let _env_lock = sys::os::env_lock();
-            cvt(libc::fork())?
-        };
-
+        // Note that as soon as we're done with the fork the parent can stop
+        // holding the lock, because it won't do anything. And the child must
+        // forget it, keeping it locked. This is because parking_lot is
+        // not fork safe and the child process must not lock or unlock any
+        // locks before it gets to libc::execvp.
         let pid = unsafe {
-            match result {
+            let _env_lock = sys::os::env_lock();
+            match cvt(libc::fork())? {
                 0 => {
+                    mem::forget(_env_lock);
```

Comment:
`parking_lot` is not fork-safe. If there were more than one thread running at the time of the fork and any of those other threads held any bucket locks inside `parking_lot` those locks will never be released, because `fork()` will just kill those threads. As such, any subsequent bucket interaction within the child process will cause a deadlock. Buckets are accessed on most lock operations, including unlocking.

As such, the child process can't be allowed to release the `env_lock`. Nor can it do anything else that requires locking between the fork and the subsequent call to `execvp`.

@Amanieu also had a theory about how we could make `parking_lot` fork safe. This included breaking the API and exposing special functions that will lock and unlock all buckets before and after the fork.